cpp/src/glm/qn/qn_solvers.cuh

```diff
@@ -41,9 +41,9 @@
  */
 
 #include <cuda_utils.h>
-#include <glm/qn/qn_linesearch.h>
-#include <glm/qn/qn_util.h>
```

Comment:
This is a bit weird. I see that `qn_util.h` does not actually have any headers that in itself is cuh (only logger.hpp)
However, I found that there is HDI, which usually stands for Host and Device inline method.
I have 2 questions here:
1) How can we use HDI here, I mean cuda_utils.h which is where HDI is defined, is not included here. If it comes from somewhere else, I still feel maybe including cuda_utils.h could help in this case.
2) Does HDI actually mean even device method? Also an extension to the question, if a method has `__host__`, while it is host method, isnt nvcc the only compiler possible to recognise this? I have seen a header file with just __host__ methods, and while I left it to be .h, I was wondering if this is right. Please correct me if I got the compilation sequence wrong (with any material that could help, even for future)

Reply:
Arghh... this might be a classic case of assuming transitive includes across header files! That is not good. :(

Q1: Yes, we should then be including cuda_utils.h inside qn_util.h and thereby renaming to `qn_util.cuh`.

Q2: yes, HDI means that method will be compiled (and thus callable) from another device methods, other cuda kernels, as well as from other host methods. Yes, for now just leave headers with only host methods to be .h

Reply:
> Yes, we should then be including cuda_utils.h inside qn_util.h and thereby renaming to qn_util.cuh.

Yes, I had moved to .cuh already, added include to cuda_utils.h now. However, I am still not sure on how the include graph goes on and how HDI define comes up.

> Yes, for now just leave headers with only host methods to be .h

Hmm, but that still leaves a question, if I use __host__ in code, would I not need `nvcc` compiler to compile?

Reply:
ok.. I guess some form of transitive include was helping qn_util.h see the definition for HDI!

Hmm.. that's a good question. I'd say give it a shot and see what happens (for eg: if that is a templated method, it *might* work). Seems like we need to further refine the rules in #1675 if this works out.